File: concrete/src/main/java/pcb/uwu/amount/base/Amperes.java
package pcb.uwu.amount.base;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.amount.derived.electromagnetism.Coulombs;
import pcb.uwu.amount.derived.electromagnetism.ElectricConductance;
import pcb.uwu.amount.derived.electromagnetism.ElectricInductance;
import pcb.uwu.amount.derived.electromagnetism.ElectricPotential;
import pcb.uwu.amount.derived.electromagnetism.ElectricResistance;
import pcb.uwu.amount.derived.electromagnetism.MagneticFlux;
import pcb.uwu.amount.derived.electromagnetism.Siemens;
import pcb.uwu.amount.derived.electromagnetism.Volts;
import pcb.uwu.amount.derived.electromagnetism.Webbers;
import pcb.uwu.amount.derived.thermodynamics.Joules;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.base.ElectricCurrentUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.base.AmpereUnit.AMPERE;
import static pcb.uwu.unit.base.SecondUnit.SECOND;
import static pcb.uwu.unit.derived.electromagnetism.HenryUnit.HENRY;
import static pcb.uwu.unit.derived.electromagnetism.OhmUnit.OHM;
import static pcb.uwu.unit.derived.electromagnetism.SiemensUnit.SIEMENS;
import static pcb.uwu.unit.derived.electromagnetism.VoltUnit.VOLT;
import static pcb.uwu.unit.derived.electromagnetism.WebberUnit.WEBBER;
import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class Amperes extends ElectricCurrent {

	// region constructors

	public Amperes(Number value) {
		super(value, AMPERE);
	}

	public Amperes(Number value, Magnitude magnitude) {
		super(value, magnitude, AMPERE);
	}

	public Amperes(String value) {
		super(value, AMPERE);
	}

	public Amperes(String value, Magnitude magnitude) {
		super(value, magnitude, AMPERE);
	}

	public Amperes(BigDecimal value) {
		super(value, AMPERE);
	}

	public Amperes(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, AMPERE);
	}

	public Amperes(BigDecimalAmount amount) {
		super(amount, AMPERE);
	}

	public Amperes(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, AMPERE);
	}

	// endregion

	// region implement UnitAmount

	@NotNull
	@Override
	public Amperes plus(@NotNull UnitAmount<ElectricCurrentUnit> other) {
		return new Amperes(plusAmount(this, other));
	}

	@NotNull
	@Override
	public Amperes minus(@NotNull UnitAmount<ElectricCurrentUnit> other) {
		return new Amperes(minusAmount(this, other));
	}

	@Override
	public Amperes times(BigDecimal other, MathContext mathContext) {
		return new Amperes(multipliedByScalar(this, other, mathContext));
	}

	@Override
	public Amperes div(BigDecimal other, MathContext mathContext) {
		return new Amperes(dividedByScalar(this, other, mathContext));
	}

	// endregion

	// region composition

	public Coulombs times(Time time, MathContext mathContext) {
		return new Coulombs(getAmount().times(getAmountIn(time, SECOND), mathContext));
	}

	public Volts times(ElectricResistance electricResistance, MathContext mathContext) {
		return new Volts(getAmount().times(getAmountIn(electricResistance, OHM), mathContext));
	}

	public Webbers times(ElectricInductance electricInductance, MathContext mathContext) {
		return new Webbers(getAmount().times(getAmountIn(electricInductance, HENRY), mathContext));
	}

	public Joules times(MagneticFlux magneticFlux, MathContext mathContext) {
		return new Joules(getAmount().times(getAmountIn(magneticFlux, WEBBER), mathContext));
	}

	public Siemens div(ElectricPotential electricPotential, MathContext mathContext) {
		return new Siemens(getAmount().div(getAmountIn(electricPotential, VOLT), mathContext));
	}

	public Volts div(ElectricConductance electricConductance, MathContext mathContext) {
		return new Volts(getAmount().div(getAmountIn(electricConductance, SIEMENS), mathContext));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/base/Amperes.kt
package pcb.uwu.amount.base

import pcb.uwu.amount.derived.electromagnetism.*
import pcb.uwu.amount.derived.thermodynamics.Joules
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.AmpereUnit
import pcb.uwu.unit.base.ElectricCurrentUnit
import pcb.uwu.unit.base.SecondUnit
import pcb.uwu.unit.derived.electromagnetism.*

class Amperes : ElectricCurrent
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, AmpereUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, AmpereUnit)

    // region UnitAmount

    override fun plus(electricCurrrent: UnitAmount<ElectricCurrentUnit>) =
        Amperes(this.amount + (electricCurrrent to AmpereUnit).amount)

    override fun minus(electricCurrrent: UnitAmount<ElectricCurrentUnit>) =
        Amperes(this.amount - (electricCurrrent to AmpereUnit).amount)

    override fun times(scalar: Number) =
        Amperes(this.amount * scalar)

    override fun div(scalar: Number) =
        Amperes(this.amount / scalar)

    // endregion

    // region composition

    operator fun times(time: Time) =
        Coulombs(this.amount * (time to SecondUnit).amount)

    operator fun times(electricResistance: ElectricResistance) =
        Volts(this.amount * (electricResistance to OhmUnit).amount)

    operator fun times(electricInductance: ElectricInductance) =
        Webbers(this.amount * (electricInductance to HenryUnit).amount)

    operator fun times(magneticFlux: MagneticFlux) =
        Joules(this.amount * (magneticFlux to WebberUnit).amount)

    operator fun div(electricPotential: ElectricPotential): Siemens
    {
        return Siemens(this.amount / (electricPotential to VoltUnit).amount)
    }

    operator fun div(electricConductance: ElectricConductance): Volts
    {
        return Volts(this.amount / (electricConductance to SiemensUnit).amount)
    }

    // endregion
}